src/core/RStream.ml

```diff
@@ -105,6 +105,17 @@ let rec map f = function
   let helper {zz} as s = {s with zz = fun () -> map f (zz ())} in
   Waiting (List.map helper ss)
 
+let mapi f =
+  let rec helper1 i = function
+  | Nil          -> Nil
+  | Cons (x,xs)  -> Cons (f i x, helper1 (1+i) xs)
+  | Thunk zzz    -> from_fun (fun () -> helper1 i @@ zzz ())
+  | Waiting ss   ->
+    let helper {zz} as s = {s with zz = fun () -> helper1 i (zz ())} in
```

Comment:
`ss` is a list of thunks (i.e. suspended streams), here you would apply `helper1` to each of them with the same initial counter equal to `i`, which I guess, can lead to an unexpected result. 
However, because `Waiting` can appear only if tabling is used somewhere in the relation, in case you don't use tabling there should be no problems. 

I currently have no idea how `Waiting` can be handled here without pain, since each of the suspended streams can have unbounded length potentially.